### ISSUE TITLE
Error in sema parsing contract

### ISSUE THREAD

**Issue by ntrippar** [hyperledger-labs/solang#731]
I found that parsing this contract ends in a stack overflow.
```solidity
pragma solidity ^0.8.7;

contract TestContract {
    struct Data {
        bytes32 a1;
        bytes32 a2;
    }

    mapping(uint176 => bytes32) Hashes;

    constructor() {}

    function getdata(uint176 id, bool needHash) public view returns (Data memory g) {
        (g.a1, g.a2) = (Hashes[id+1], Hashes[id+2]);
    }
}
```

this seems to be when its casting the `StorageRef(false, Bytes(32))` to  `Ref(Bytes(32))`

```js
from cast: StorageRef(false, Bytes(32))
to cast: Ref(Bytes(32))
from cast: Ref(StorageRef(false, Bytes(32)))
to cast: Ref(Bytes(32))
```
based on the result that I was getting with some debug prints in the function `cast` I noticed that the function checks if the `to` needs a `Ref`, and try creates a `Ref` from the `from`.  this cause an infinite loop.

https://github.com/hyperledger-labs/solang/blob/5da5377b6fce0d1aec98fa58041a4126540a226d/src/sema/expression.rs#L540-L550

one solution that seems to be working is not dereference when the `to` object type is a `Type::StorageRef`, but as I don't have experience with this codebase I didn't create a pull request.

```rust
  // First of all, if we need a ref then derefence it
  if matches!(to, Type::Ref(..)) && !matches!(from, Type::Ref(..)) && !matches!(from, Type::StorageRef(..)) {
      return cast(
          loc,
          Expression::GetRef(*loc, Type::Ref(Box::new(from)), Box::new(expr)),
          to,
          implicit,
          ns,
          diagnostics,
      );
  }
```

**Comment by ntrippar** [comment]
Also do the project has a "repo" with a lot of testcases?, if not I would recommend that the project should have one with a lot of sol files named `{github_issue_id}.sol` an a `{github_issue_id}.json` with some information that can do continuous test to check if there is any regression of a bug.  or something in broken

**Comment by seanyoung** [comment]
> Also do the project has a "repo" with a lot of testcases?, if not I would recommend that the project should have one with a lot of sol files named `{github_issue_id}.sol` an a `{github_issue_id}.json` with some information that can do continuous test to check if there is any regression of a bug. or something in broken

There are a whole bunch of testcases in the tests/ directories, including over 500 sol files.

**Comment by seanyoung** [comment]
Looks like the casting is broken in this situation. That bit of code is a bit ugly, it might need some re-work. We will add this example as a testcase once fixed

**Comment by ntrippar** [comment]
> > Also do the project has a "repo" with a lot of testcases?, if not I would recommend that the project should have one with a lot of sol files named `{github_issue_id}.sol` an a `{github_issue_id}.json` with some information that can do continuous test to check if there is any regression of a bug. or something in broken
> 
> There are a whole bunch of testcases in the tests/ directories, including over 500 sol files.

ah great, I will fill some more issues with other bugs that I found using solang.

**Comment by ntrippar** [comment]
also I think this following case also is wrong as end also in a stack overflow

```rust
to: Ref(Uint(256)), from: Uint(112)
to: Ref(Uint(256)), from: Ref(Uint(112))
to: Ref(Uint(256)), from: Uint(112)
to: Ref(Uint(256)), from: Ref(Uint(112))
to: Ref(Uint(256)), from: Uint(112)
to: Ref(Uint(256)), from: Ref(Uint(112))
```

https://github.com/hyperledger-labs/solang/blob/5da5377b6fce0d1aec98fa58041a4126540a226d/src/sema/expression.rs#L552-L562

and should be something like 

```rs
// First of all, if we have a ref then derefence it
if matches!(from, Type::Ref(..)) && !matches!(to, Type::Ref(..)){
    if let Type::Ref(r) = from {
        return cast(
            loc,
            Expression::Load(*loc, r.as_ref().clone(), Box::new(expr)),
            to,
            implicit,
            ns,
            diagnostics,
        );
    }
}
```


**Testcase**

```solidity
pragma solidity 0.6.12;

interface IUniswapV2Pair {
    function getReserves() external view returns (uint112 reserve0, uint112 reserve1, uint32 blockTimestampLast);   
}


contract Contract {
   
    struct Struct1 {
        uint256 a;
        uint256 b;
    }

    function test(address[] memory _tokens) public view {

        // get shares and eth required for each share
        Struct1[] memory struct_1 = new Struct1[](3);

        (struct_1[0].a, struct_1[0].b,) = IUniswapV2Pair(_tokens[0]).getReserves();

    }
}
```

**Comment by seanyoung** [comment]
Fixed in #738
